sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -568,8 +574,13 @@ object DataSource extends Logging {
     "org.apache.spark.Logging")
 
   /** Given a provider name, look up the data source class definition. */
-  def lookupDataSource(provider: String): Class[_] = {
-    val provider1 = backwardCompatibilityMap.getOrElse(provider, provider)
+  def lookupDataSource(provider: String, conf: SQLConf): Class[_] = {
+    val provider1 = backwardCompatibilityMap.getOrElse(provider, provider) match {
+      case name if name.equalsIgnoreCase("orc") &&
+          conf.getConf(SQLConf.ORC_IMPLEMENTATION) == "native" =>
+        classOf[OrcFileFormat].getCanonicalName
+      case name => name
```

Comment:
if `ORC_IMPLEMENTATION` is `hive`, we leave the provider as it was, which may be `orc`. Then we will hit `Multiple sources found` issue, aren't we? Both the old and new orc has the same short name `orc`.

Reply:
I was looking the exact same path. It seems not because  it's not registered to `ServiceLoader` (`src/main/resources/org.apache.spark.sql.sources.DataSourceRegister`). So, short name for the newer ORC source would not find be used.

Reply:
@cloud-fan . To avoid that issue, new OrcFileFormat is not registered intentionally.
@HyukjinKwon 's comment is correct.

Reply:
This sounds counter-intuitive, I think we should register the new orc instead of the old one.

Reply:
and also add comments here.

Reply:
+1 for ^

Reply:
I agree with both of you.

Just for explanation: The original design completely preserves the previous behavior.
Without `SQLConf.ORC_IMPLEMENTATION` option, Spark doesn't know OrcFileFormat. So, in case of non-Hive support, creating data source with "orc" will fail with unknown data source.

Anyway, I'm happy to update according to your advice. :)

Reply:
So, there is no more `The ORC data source must be used with Hive support enabled`.

Reply:
sounds good

Reply:
And for here, I added the following to prevent `Multiple sources found`. Last time, I missed this way. My bad.
```
+      case name if name.equalsIgnoreCase("orc") &&
+        conf.getConf(SQLConf.ORC_IMPLEMENTATION) == "hive" =>
+        "org.apache.spark.sql.hive.orc.OrcFileFormat"
```